### PR TITLE
[1/n] upgrade react-native 0.73

### PR DESCRIPTION
**Pull request by gabrieldonadel**
# Why

upgrade react-native 0.73 for sdk 50
Closes ENG-9739

# How

- update package versions
  - `react-native 0.72.5 -> 0.73.0-rc.4` 
  - `@react-native/assets-registry 0.72.0 -> 0.73.1` 
  - `metro-react-native-babel-preset 0.76.8 -> @react-native/babel-preset 0.73.18`
- upgrade project templates based on [upgrade-helper](https://react-native-community.github.io/upgrade-helper/?from=0.72.5&to=0.73.0-rc.2)
- Backport folly_version bump to v2022.05.16.00
- Expo Go
  - Add Agp configurator plugin in order to make sure `buildConfig` is turned on for all the 3rd party libraries and ensure namespace is specified for all the 3rd party libraries
  - Migrate go to version catalogs 
- Temporarily patched @react-native/dev-middleware while `react-native@0.73.0-rc.5` is not out (check https://github.com/facebook/react-native/pull/41370 for more context)
- cli 
  - Migrate metro-inspector-proxy to @react-native/dev-middleware
  - remove runInspectorProxy logic
- [gesture-handler] Temporarily patched patch before https://github.com/software-mansion/react-native-gesture-handler/commit/4efaebce87d3f2949d72f477ec896b03e95f90f4 is released 
- [modules][dev-menu] Update import paths for fabric
- [html-elements] Update Style types
- for other details, please check the commit histories one by one. 

# Next steps

Part 2 -> https://github.com/expo/expo/pull/25453
- Upgrade react-native to 0.73.0 when the official release is out
- Remove all patches 
- Part 2
  - Move AgpConfiguratorPlugin to a different file
  - Migrate MainApplication / MainActivity to kotlin

# Test Plan

- bare-expo ios / android
- Expo Go
- fabric ios / android
- ci passed

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: package.json
Comment (gabrieldonadel) [inline package.json:10]: Temporary patch while https://github.com/facebook/react-native/pull/41370 is not released

File: android/versioned-abis/expoview-abi49_0_0/src/main/java/abi49_0_0/expo/modules/splashscreen/NativeResourcesBasedSplashScreenViewProvider.kt
Comment (Kudo) [inline android/versioned-abis/expoview-abi49_0_0/src/main/java/abi49_0_0/expo/modules/splashscreen/NativeResourcesBasedSplashScreenViewProvider.kt:10]: just a note that may break versioning.

File: android/versioned-abis/expoview-abi49_0_0/src/main/java/abi49_0_0/expo/modules/splashscreen/SplashScreenReactActivityLifecycleListener.kt
Comment (Kudo) [inline android/versioned-abis/expoview-abi49_0_0/src/main/java/abi49_0_0/expo/modules/splashscreen/SplashScreenReactActivityLifecycleListener.kt:12]: just a note that may break versioning.

File: android/versioned-abis/expoview-abi49_0_0/src/main/java/abi49_0_0/expo/modules/systemui/SystemUIReactActivityLifecycleListener.kt
Comment (Kudo) [inline android/versioned-abis/expoview-abi49_0_0/src/main/java/abi49_0_0/expo/modules/systemui/SystemUIReactActivityLifecycleListener.kt:9]: just a note that may break versioning.

File: apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainApplication.java
Comment (Kudo) [inline apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainApplication.java:13]: nit: do you want to migrate MainApplication / MainActivity to kotlin as 0.73?

File: apps/native-component-list/metro.config.js
Comment (Kudo) [inline apps/native-component-list/metro.config.js:65]: could you share any insight for removing this?

File: packages/expo-modules-core/common/cpp/fabric/ExpoViewComponentDescriptor.cpp
Comment (Kudo) [inline packages/expo-modules-core/common/cpp/fabric/ExpoViewComponentDescriptor.cpp:9]: that's interesting to me. we had 0.73 on react-native-modules but not seeing this error before. would take a look further.

**Review by Kudo — COMMENTED** [review]
it must be tough for all the breaking changes, but you did a great work.
i just ran first round of review and left some feedback.

File: packages/expo-modules-core/common/cpp/fabric/ExpoViewComponentDescriptor.cpp
Comment (gabrieldonadel) [reply packages/expo-modules-core/common/cpp/fabric/ExpoViewComponentDescriptor.cpp:9]: Yeah, I'm not sure why this is happening but Reanimated is facing a similar problem 

https://github.com/expo/expo/blob/70bee13f4e10a09d20f394f689a3b9c5a8dc3dd3/ios/vendored/unversioned/react-native-reanimated/Common/cpp/NativeModules/NativeReanimatedModule.cpp#L39

File: android/build.gradle
Comment (gabrieldonadel) [reply android/build.gradle:121]: Exactly

File: apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainApplication.java
Comment (gabrieldonadel) [reply apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainApplication.java:13]: Sure, I can migrate this as the part 2 of this PR

File: packages/expo-modules-core/common/cpp/fabric/ExpoViewComponentDescriptor.cpp
Comment (Kudo) [reply packages/expo-modules-core/common/cpp/fabric/ExpoViewComponentDescriptor.cpp:9]: ok for this one. the correct fix should be
```diff
--- a/packages/expo-modules-core/android/src/fabric/CMakeLists.txt
+++ b/packages/expo-modules-core/android/src/fabric/CMakeLists.txt
@@ -43,6 +43,7 @@ target_link_libraries(fabric
   ReactAndroid::react_render_debug
   ReactAndroid::react_render_graphics
   ReactAndroid::react_render_mapbuffer
+  ReactAndroid::react_utils
   ReactAndroid::rrc_view
   ReactAndroid::runtimeexecutor
   ReactAndroid::yoga
```

File: yarn.lock
Comment (Kudo) [inline yarn.lock:16672]: there's another missing piece react-native@0.72.5 at https://github.com/expo/expo/blob/a63166684aa2580af9ac34a36143e424fa1bf486/apps/sandbox/package.json#L14
that's why the lock file here still have 0.72.5

Comment (gabrieldonadel) [reply yarn.lock]: Sorry, missed that. updated

File: apps/native-component-list/metro.config.js
Comment (Kudo) [reply apps/native-component-list/metro.config.js:65]: checked the bundle without the `getModulesRunBeforeMainModule` here and it includes the necessary imports. it looks like we don't need the workaround here anymore 🎉
context: that is how i use wget to check the bundle #20992

File: android/build.gradle
Comment (Kudo) [reply android/build.gradle:121]: looks like another AGP 8 breaking changes.
however, i think adding pickFirsts everywhere is not a perfect idea. thank Doug for having #25334, now we ran android instrumented test on bare-expo. i think we could just remove the entire block here.
https://github.com/expo/expo/blob/5204e45053c19ba8f7461103960e38462f948190/android/build.gradle#L111-L124

Comment (gabrieldonadel) [reply android/build.gradle]: removed

File: android/expoview/build.gradle
Comment (Kudo) [reply android/expoview/build.gradle:236]: oh right. its worth to try out having another version catalog and reading from react-native's toml file https://docs.gradle.org/current/userguide/platforms.html#sec:importing-catalog-from-file

File: packages/expo-modules-core/common/cpp/fabric/ExpoViewComponentDescriptor.cpp
Comment (gabrieldonadel) [reply packages/expo-modules-core/common/cpp/fabric/ExpoViewComponentDescriptor.cpp]: Ohh thanks for finding a proper fix for this, updated

File: android/app/build.gradle
Comment (gabrieldonadel) [reply android/app/build.gradle:158]: Added the comment for now just to unlock this so we can start upgrading modules soon, but I can address this in part 2

File: android/tools/src/main/java/host/exp/exponent/tools/ReactAndroidCodeTransformer.java
Comment (gabrieldonadel) [reply android/tools/src/main/java/host/exp/exponent/tools/ReactAndroidCodeTransformer.java:315]: I agree with this, the ReactAndroidCodeTransformer seems a bit fragile and I can move these to commits in the fork, I would like to do this along with the other adjustments on a follow up PR tho, in order to unblock the upgrade of the other modules

File: android/expoview/build.gradle
Comment (gabrieldonadel) [reply android/expoview/build.gradle:236]: Sure, I've updated this to import the version catalog from react-native-labs

File: apps/bare-expo/e2e/TestSuite-test.native.js
Comment (gabrieldonadel) [inline apps/bare-expo/e2e/TestSuite-test.native.js:16]: Reenabled here https://github.com/expo/expo/pull/25453

File: apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainApplication.java
Comment (gabrieldonadel) [reply apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainApplication.java:13]: will be addressed here https://github.com/expo/expo/pull/25453

File: packages/expo-updates/android/src/androidTest/java/expo/modules/updates/logging/UpdatesLoggingTest.kt
Comment (gabrieldonadel) [inline packages/expo-updates/android/src/androidTest/java/expo/modules/updates/logging/UpdatesLoggingTest.kt:128]: Synced with @douglowder on Slack about this, Will will investigate after this lands on `main`

File: android/libs.versions.toml
Comment (Kudo) [inline android/libs.versions.toml:27]: nit: can remove some libs here because we read the versions from toml inside react-native-lab.

File: apps/native-tests/ios/NativeTests.xcodeproj/project.pbxproj
Comment (Kudo) [inline apps/native-tests/ios/NativeTests.xcodeproj/project.pbxproj:367]: nit: remove the workaround

File: apps/native-tests/ios/NativeTests.xcodeproj/project.pbxproj
Comment (Kudo) [inline apps/native-tests/ios/NativeTests.xcodeproj/project.pbxproj:425]: remove the workaround

File: ios/versioned-react-native/ABI49_0_0/ReactNative/React/ABI49_0_0React-RCTFabric.podspec
Comment (Kudo) [inline ios/versioned-react-native/ABI49_0_0/ReactNative/React/ABI49_0_0React-RCTFabric.podspec:13]:
```suggestion
folly_flags = '-DFOLLY_NO_CONFIG -DFOLLY_MOBILE=1 -DFOLLY_USE_LIBCPP=1 -DFOLLY_CFG_NO_COROUTINES=1'
```

sorry i missed this one

File: ios/versioned-react-native/ABI49_0_0/ReactNative/ReactCommon/hermes/ABI49_0_0React-hermes.podspec
Comment (Kudo) [inline ios/versioned-react-native/ABI49_0_0/ReactNative/ReactCommon/hermes/ABI49_0_0React-hermes.podspec:17]:
```suggestion
folly_compiler_flags = '-DFOLLY_NO_CONFIG -DFOLLY_MOBILE=1 -DFOLLY_USE_LIBCPP=1 -DFOLLY_CFG_NO_COROUTINES=1 -DFOLLY_HAVE_CLOCK_GETTIME=1 -Wno-comma -Wno-shorten-64-to-32'
```

sorry i missed this one

**Review by Kudo — APPROVED** [review]
thanks for the great work to make all ci green. leaving some comments and you could follow up in pr part 2 for sure.

File: package.json
Comment (tsapeta) [inline package.json:10]: Might be a good idea to move this to a separate script file 🤔

**Review by tsapeta — APPROVED** [review]
Oh wow, you did so much work! 😮

File: ios/versioned-react-native/ABI49_0_0/ReactNative/React/ABI49_0_0React-RCTFabric.podspec
Comment (gabrieldonadel) [reply ios/versioned-react-native/ABI49_0_0/ReactNative/React/ABI49_0_0React-RCTFabric.podspec]: Updated

File: ios/versioned-react-native/ABI49_0_0/ReactNative/ReactCommon/hermes/ABI49_0_0React-hermes.podspec
Comment (gabrieldonadel) [reply ios/versioned-react-native/ABI49_0_0/ReactNative/ReactCommon/hermes/ABI49_0_0React-hermes.podspec]: Updated

File: apps/native-tests/ios/NativeTests.xcodeproj/project.pbxproj
Comment (gabrieldonadel) [reply apps/native-tests/ios/NativeTests.xcodeproj/project.pbxproj]: removed

File: apps/native-tests/ios/NativeTests.xcodeproj/project.pbxproj
Comment (gabrieldonadel) [reply apps/native-tests/ios/NativeTests.xcodeproj/project.pbxproj]: removed

File: package.json
Comment (gabrieldonadel) [reply package.json:10]: maybe we should use patch-package?

Comment (Kudo) [reply package.json:10]: IIRC, patch-package does not work (maybe due to our monorepo setup). @alanjhughes had some problem with that before.
furthermore, i would prefer not add new dependencies to our repo just for the temporary fixes. we will soon to start updating the vendor modules and should be able to remove these patches then.

File: android/libs.versions.toml
Comment (gabrieldonadel) [reply android/libs.versions.toml:15]: Removed

**Pull request merged**
